tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,524 @@
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        let res = aks.add_keys(testname, keys.clone(), is_replace, is_force);
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Err(err.into())
+            }
+        }
+    }
+
+    // A wrapper for deleting an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like valid1, bad
+    // testname: key name given as cmdline args, like "one", "two"
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn del_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str) {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        for key in aks.remove_keys(testname) {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                aks.write().expect("failed to update authorized keys directory");
+                aks.sync().expect("failed to update authorized keys");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+                assert!(!authkeyone.exists());
+
+                let mut svec: Vec<&str> = Vec::new();
+                svec.push(pubkeyname);
+                assert_has_keys(ssh_dir, svec, true);
+            }
+        }
+    }
+
+    fn setup_tests() -> TestContext {
+        let ssh_dir_path: PathBuf = TempDir::new(SSH_DIR_PREFIX).unwrap().into_path();
+
+        for (name, text) in &*TEST_KEYS {
+            let pub_path = ssh_dir_path.join(format!("{}.pub", name));
+
+            let mut pubfile = File::create(pub_path.clone()).expect(format!("unable to create a file {}", pub_path.to_str().unwrap_or_default()).as_str());
+            let _ = pubfile.write_all(text.as_bytes());
+        }
+
+        test_no_keys(ssh_dir_path.to_str().unwrap());
+        test_first_run(ssh_dir_path.to_str().unwrap());
+
```

Comment:
Please remove this here, the execution of this gives a different meaning to all tests using this function because the state is already modified then. Both tests should use their own `#[test]` function.

Reply:
I already went down to that route, but the whole test did not work, like:

```
thread 'tests::test_no_replace' panicked at 'failed to update authorized keys: Error(NoKeysFound("\"/tmp/test_update_ssh_keys.UrqDaXpWT31V/authorized_keys.d\""), State { next_error: None, backtrace: InternalBacktrace })'
, tests/test_update_ssh_keys.rs:235:20 
```

Tests passed only when a single test runs, or all tests run in only 1 thread.

Reply:
Then something else is not right, but invoking other tests before each tests and they leave their state is not the solution: The tests should keep their original semantics.

Reply:
I found the problem: `add_one_ssh_key` did not sync and write itself but relied on older state.